### PR TITLE
Fix critical port 0 dynamic assignment bugs and enhance devtools API

### PR DESCRIPTION
**Pull request by elear-solutions-dev**
This commit fixes two critical bugs in MeshLink's port handling and enhances the devtools API for better testing capabilities.

BUG FIXES:

1. Fix mesh->myport remaining "0" after OS assigns ephemeral port
   - Problem: When meshlink_set_port(mesh, 0) is called, the OS assigns an actual ephemeral port, but mesh->myport remains "0"
   - Impact: REQ_EXTERNAL messages advertise port 0, breaking P2P connections
   - Fix: Update mesh->myport with actual assigned port using getsockname()
   - Files: src/net_setup.c, src/meshlink.c

2. Fix meshlink_set_port() clearing external_ip_address
   - Problem: meshlink_set_port() after meshlink_open() clears external_ip_address
   - Impact: Backbone nodes can't send REQ_EXTERNAL messages, breaking P2P
   - Fix: Preserve external_ip_address before network reinit, restore after
   - Files: src/meshlink.c

3. Fix misleading "Listening on port 0" log messages
   - Problem: Log messages showed port 0 even after OS assigned actual port
   - Fix: Update socket address structures and move logging after port update
   - Files: src/net_setup.c

ENHANCEMENTS:

4. Extend devtools API for comprehensive testing
   - Add external_ip_address and canonical_address fields to devtool_node_status_t
   - Add devtool_free_node_status() for proper memory management
   - Export new function in meshlink.sym
   - Files: src/devtools.h, src/devtools.c, src/meshlink.sym

5. Enhanced test coverage
   - Extend port.c test for port 0 dynamic assignment and port conflicts
   - Add req-external-port.c test for REQ_EXTERNAL message validation
   - Test both BACKBONE+BACKBONE and BACKBONE+PORTABLE scenarios
   - Files: test/port.c, test/req-external-port.c, test/Makefile.am

TECHNICAL DETAILS:

- Uses getsockname() to retrieve actual OS-assigned port from bound socket
- Preserves external IP address during port changes to maintain P2P connectivity
- Updates all socket address structures to reflect actual assigned ports
- Implements proper memory management for devtools API extensions
- Comprehensive test coverage for both bugs and edge cases

All tests pass and the fixes resolve the critical P2P connection issues caused by incorrect port advertisement in REQ_EXTERNAL messages.

---

> [!NOTE]
> Fixes port 0 handling (assign, persist, and advertise actual port; preserve external_ip on port change), adds devtools fields for external/canonical addresses with cleanup, and extends tests for port/REQ_EXTERNAL behavior.
> 
> - **Networking (port handling)**:
>   - Update `add_listen_sockets()` to replace `mesh->myport` when binding to `port 0` using `getsockname()`, and log after update in `src/net_setup.c`.
>   - In `meshlink_set_port()` (`src/meshlink.c`): preserve and restore `self->external_ip_address` across network reinit; treat `port==0` as success when an actual port is assigned; guard recent-address rebuild.
>   - Header docs updated in `meshlink.h` to clarify `meshlink_get_port()` and `meshlink_set_port()` behavior with `port 0`.
> - **Devtools API**:
>   - Extend `devtool_node_status_t` with `external_ip_address` and `canonical_address`; duplicate values in `devtool_get_node_status()` and add `devtool_free_node_status()`; export symbol in `meshlink.sym`.
> - **Tests**:
>   - Enhance `test/port.c` to verify dynamic `port 0` assignment, persistence, and conflict handling.
>   - Add `test/req-external-port.c` to validate REQ_EXTERNAL learning (backbone↔backbone and backbone→portable).
>   - Wire new test into `test/Makefile.am`; ignore test binaries in `.gitignore`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d0473fa64f79b817c7f90e1796eb868d314dad07. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by elear-solutions-dev** [comment]
Bugbot run

**Comment by elear-solutions-dev** [comment]
Bugbot run

File: test/req-external-port.c [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	printf("Starting mesh2 (portable)...\n");
+	assert(meshlink_start(mesh2));
+
+	// Wait for connection
+	printf("Waiting for nodes to connect...\n");
+	wait_sync_flag(&bar_reachable, 5);
+
+	printf("✅ Nodes connected successfully\n");
```
[the remark this answers is not in the document]
Comment (elear-solutions-dev) [reply test/req-external-port.c:228]: Fixed

File: test/port [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  absdir=`cd "$thisdir" && pwd`
+  test -n "$absdir" && thisdir="$absdir"
+
+  program='port'
+  progdir="$thisdir/.libs"
+
+
+  if test -f "$progdir/$program"; then
+    # Add our own library path to LD_LIBRARY_PATH
+    LD_LIBRARY_PATH="/mnt/host/coco-project/meshlink-downstream/src/.libs:$LD_LIBRARY_PATH"
```
[the remark this answers is not in the document]
Comment (elear-solutions-dev) [reply test/port]: Fixed by removing the libtool file. This shouldn't have been part of the PR.

**Comment by elear-solutions-dev** [comment]
Bugbot run

File: src/meshlink.c [hunk 3]
```diff
@@ -2976,6 +2977,16 @@ bool meshlink_set_port(meshlink_handle_t *mesh, int port) {
...
+	// BUG FIX: Preserve external_ip_address before closing network connections
+	// When meshlink_set_port() is called after meshlink_open(), the external_ip_address
+	// gets lost because close_network_connections() deletes mesh->self. This prevents
+	// REQ_EXTERNAL messages from being sent, breaking P2P connectivity.
+	// We preserve the IP part (without port) and will restore it with the new port.
+	if(mesh->self && mesh->self->external_ip_address) {
+		saved_external_ip = xstrdup(mesh->self->external_ip_address);
+		logger(mesh, MESHLINK_DEBUG, "Preserving external IP address: %s", saved_external_ip);
+	}
+
```
[the remark this answers is not in the document]
Comment (elear-solutions-dev) [reply src/meshlink.c:2989]: There was no memory leak - the second allocation is necessary because close_network_connections() freed the first allocation. The original code is working correctly, right? Raise another comment if you disagree.

**Comment by elear-solutions-dev** [comment]
Bugbot run

File: test/req-external-port.c [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	printf("✅ Verifying REQ_EXTERNAL bidirectional learning...\n");
+	
+	// Get nodes for testing
+	meshlink_node_t *node_foo = meshlink_get_node(mesh2, "foo");
+	meshlink_node_t *node_bar = meshlink_get_node(mesh1, "bar");
+	assert(node_foo && node_bar);
+	
+	// Test bidirectional learning: bar should learn foo's external info
+	devtool_node_status_t bar_view_of_foo;
+	devtool_get_node_status(mesh1, node_foo, &bar_view_of_foo);
```
[the remark this answers is not in the document]
Comment (elear-solutions-dev) [reply test/req-external-port.c]: Excellent catch. Fixed

**Comment by elear-solutions-dev** [comment]
Bugbot run
